### PR TITLE
added select item to draw

### PR DESCRIPTION
**Pull request by YanivDash**
### added or fixed these functionalities

1. draw an image on the mouse click 
2.  prevent image drawing when clicked on boxes
3.  select an image to draw 
4. files explorer -> will return all items at all levels (currently not traversing deeper than one level)

### CODE REVIEW

File: http_server/FileExplorer.js [hunk 1]
```diff
@@ -4,28 +4,34 @@ const path = require('path')
...
-			}
-			else {
-				const folderName = path.dirname(file)
-				if (!filesAndFolders[folderName]) {
-					filesAndFolders[folderName] = []
-				}
-				filesAndFolders[folderName].push(path.basename(file))
-			}
-		})
+  static processDirectory(directoryPath, filesAndImages) {
```
Comment (kristus123) [inline http_server/FileExplorer.js:14]: no need to extrac this into a method

File: static/engine/DynamicHtml.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
+export class DynamicHtml {
```
Comment (kristus123) [inline static/engine/DynamicHtml.js:1]: this is a bad name

File: static/engine/DynamicHtml.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+  static selectedImage = "";
+
+  static handleSelectElement(data) {
```
Comment (kristus123) [inline static/engine/DynamicHtml.js:5]: name this better as well

File: static/engine/Overlay.js [hunk 4]
```diff
@@ -14,12 +14,15 @@ function fetchHTML() {
...
 
+
 let offsetX = 0
 let offsetY = 0
 
 export class Overlay {
 	static create(camera) {
 		document.getElementById('overlay').innerHTML = html
+		let data = Http.get("/canvas-image");
+		DynamicHtml.handleSelectElement(data);
```
Comment (kristus123) [inline static/engine/Overlay.js:25]: just do the call within the class

File: static/engine/controller/mouse/ClickEvents.js [hunk 5]
```diff
@@ -11,13 +11,13 @@ export class ClickEvents {
...
 		}
 
-		const wrapperHandler = (e) => {
-			const mousePosition = this.mouse.positionRelativeToCamera(e)
-
-			if (!this.mouse.holding) {
-				handler(mousePosition)
-			}
-		}
+    const wrapperHandler = (e) => {
```
Comment (kristus123) [inline static/engine/controller/mouse/ClickEvents.js:14]: run the linter

Comment (kristus123) [reply static/engine/controller/mouse/ClickEvents.js:14]: unless this is how the linter wants it to be

File: http_server/main.js [hunk 6]
```diff
@@ -35,6 +35,20 @@ app.get('/world-editor', (req, res) => {
 	})
 })
 
+app.get("/canvas-image", (req, res) => {
+  try {
+    const data = Files.inFolder("static/assets")
```
Comment (kristus123) [inline http_server/main.js:40]: remove all error handling. no need for it in this project

File: static/engine/DynamicHtml.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+		  if (selectedValue == "root") {
+			  DynamicHtml.createImageOption(data.root, data.path);
+		  } else {
+			  data[selectedValue].path = data.path + "/" + selectedValue;
+			  DynamicHtml.handleSelectElement(data[selectedValue]);
+		  }
+	  });
+  }
+
+  static handleImageOption(data, path) {
```
Comment (kristus123) [inline static/engine/DynamicHtml.js:28]: what is data?
make data a better variable name

File: static/engine/DynamicHtml.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	const select = createElement("select", ".footer .item");
+	  data.forEach((e) => {
+		  let element = document.createElement("option");
+		  element.innerHTML = e;
+		  select.appendChild(element);
+	  });
+	  select.addEventListener("change", function () {
+		  var selectedValue = this.value;
```
Comment (kristus123) [inline static/engine/DynamicHtml.js:36]: no need for a variable here.

dont' use 'var'. that is outdated

**Review by kristus123 — CHANGES_REQUESTED** [review]

File: http_server/FileExplorer.js (hunk 1 above)
Comment (YanivDash) [reply http_server/FileExplorer.js]: done ,

File: static/engine/core/index.js [hunk 9]
```diff
@@ -15,10 +15,12 @@ ErrorHandler.run(() => {
...
 
 	const levelSelector = new LevelSelector()
-	levelSelector.changeActiveLevel(new World(levelSelector, allGameObjects, camera, mouse, controller))
-	// levelSelector.changeActiveLevel(new WorldEditor(camera, mouse))
+	//levelSelector.changeActiveLevel(new World(levelSelector, allGameObjects, camera, mouse, controller))
+	levelSelector.changeActiveLevel(new WorldEditor(camera, mouse))
 
-	// Overlay.create()
+	Overlay.create(onload =  () => {
+		ImageSelectorProvider.addImageSelector()
```
Comment (kristus123) [inline static/engine/core/index.js:22]: don't do it this way. this is just unnecessarily complicated

File: http_server/Files.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
+const fs = require("fs");
+const path = require("path");
+
+// module.exports =
+
+class Files {
```
Comment (kristus123) [inline http_server/Files.js:6]: revert files.js to how it was originally, unless there is a reason you haven't done so yet

File: static/engine/ImageSelectorProvider.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+      });
+
+      // Handle image loading error
+      img.onerror = function () {
+        imgDiv.remove();
+      };
+    });
+  }
+
+  static #createHtmlElement = (elem, parent, cls = "") => {
```
Comment (kristus123) [inline static/engine/ImageSelectorProvider.js:51]: move these into a class of  its own

File: static/engine/ImageSelectorProvider.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    const newElement = document.createElement(elem);
+
+    newElement.className = cls;
+    parentElement.appendChild(newElement);
+    return newElement;
+  };
+
+
+  // remove existing unwanted elements
+  static #removeElements(className) {
```
Comment (kristus123) [inline static/engine/ImageSelectorProvider.js:63]: move these into a class of its own

File: static/engine/Overlay.js [hunk 13]
```diff
@@ -14,12 +14,14 @@ function fetchHTML() {
 
 const html = fetchHTML()
 
+
 let offsetX = 0
 let offsetY = 0
 
 export class Overlay {
-	static create(camera) {
+	static create(camera, onLoad = () => {}) {
```
Comment (kristus123) [inline static/engine/Overlay.js:22]: don't do it like this

File: static/engine/WorldEditor.js [hunk 14]
```diff
@@ -14,14 +14,26 @@ export class WorldEditor {
...
-		})
+		mouse.addOnClick("paint", (p, c) => {
+			if (
+				c.contains("item") ||
+				c.contains("selector") ||
+				c.contains("side-item") ||
+				!ImageSelectorProvider.selectedImage
+			) {
+			} else {
+				this.selectedImage = `../${ImageSelectorProvider.selectedImage}`;
```
Comment (kristus123) [inline static/engine/WorldEditor.js:25]: why do you use relative paths here ? stay away from using relative paths

Comment (kristus123) [reply static/engine/WorldEditor.js:25]: this is just a hack to make it work

File: static/engine/WorldEditor.js [hunk 15]
```diff
@@ -14,14 +14,26 @@ export class WorldEditor {
...
+
+				p.height = this.selectedImage.width;
+				p.width = this.selectedImage.height;
+				const o = new StaticPicture(
+					new Position(p.x, p.y, p.width, p.height),
```
Comment (kristus123) [inline static/engine/WorldEditor.js:30]: you can just pass p

**Review by kristus123 — CHANGES_REQUESTED** [review]

File: static/engine/WorldEditor.js [hunk 16]
```diff
@@ -14,14 +14,26 @@ export class WorldEditor {
...
 
-		mouse.addOnClick('paint', p => {
-			p.height = 100
-			p.width = 100
-			const o = new DynamicGameObject(new Position(p.x, p.y, p.width, p.height), 100, 100)
-
-			this.runAll.add(o)
-			ObjectPersistence.save(o)
-		})
+		mouse.addOnClick("paint", (p, c) => {
```
Comment (kristus123) [inline static/engine/WorldEditor.js:17]: don't do it like this. This is not a good solution

For now just don't worry about it drawing a picture when clicking a div

File: static/engine/controller/mouse/ClickEvents.js [hunk 17]
```diff
@@ -11,13 +11,13 @@ export class ClickEvents {
...
+      const mousePosition = this.mouse.positionRelativeToCamera(e);
+      const elementClickedClass = e.target.classList;
+      if (!this.mouse.holding) {
```
Comment (kristus123) [inline static/engine/controller/mouse/ClickEvents.js:17]: just revert this

**Review by kristus123 — CHANGES_REQUESTED** [review]

File: static/engine/WorldEditor.js [hunk 18]
```diff
@@ -14,14 +14,34 @@ export class WorldEditor {
 			this.runAll.add(o)
 		})
 
-		mouse.addOnClick('paint', p => {
-			p.height = 100
-			p.width = 100
-			const o = new DynamicGameObject(new Position(p.x, p.y, p.width, p.height) ,100, 100)
+		mouse.addOnClick("paint", async (p, c) => {
```
[the remark this answers is not in the document]
Comment (YanivDash) [reply static/engine/WorldEditor.js]: remove

File: static/engine/core/index.js (hunk 9 above)
Comment (YanivDash) [reply static/engine/core/index.js:22]: i did it this way , coz the image selector element has to be loaded , only after html is loaded in overlay.create() 
so i added a callback hook parameter that will be called after the inner html loaded . 
and we also need the selected element information in world editor . 
this way it is more loosely coupled from either of the class.

Comment (kristus123) [reply static/engine/core/index.js:22]: do it the simpler way. no need to overcomplicate things

File: static/engine/controller/mouse/ClickEvents.js (hunk 17 above)
Comment (YanivDash) [reply static/engine/controller/mouse/ClickEvents.js:16]: done

File: http_server/Files.js (hunk 10 above)
Comment (YanivDash) [reply http_server/Files.js:5]: done and files will return the path now

File: static/engine/core/index.js (hunk 9 above)
Comment (YanivDash) [reply static/engine/core/index.js]: done

File: static/engine/ImageSelectorProvider.js (hunk 11 above)
Comment (YanivDash) [reply static/engine/ImageSelectorProvider.js]: done

File: http_server/main.js (hunk 6 above)
Comment (YanivDash) [reply http_server/main.js]: done

File: static/engine/DynamicHtml.js (hunk 7 above)
Comment (YanivDash) [reply static/engine/DynamicHtml.js]: changes to imageData

File: static/engine/Overlay.js (hunk 4 above)
Comment (YanivDash) [reply static/engine/Overlay.js]: done

File: http_server/main.js [hunk 19]
```diff
@@ -35,6 +35,20 @@ app.get('/world-editor', (req, res) => {
 	})
 })
 
+app.get("/canvas-image", (req, res) => {
```
[the remark this answers is not in the document]
Comment (YanivDash) [reply http_server/main.js]: done

File: static/engine/DynamicHtml.js (hunk 2 above)
Comment (YanivDash) [reply static/engine/DynamicHtml.js]: changed

File: static/engine/DynamicHtml.js (hunk 3 above)
Comment (YanivDash) [reply static/engine/DynamicHtml.js]: done

**Pull request merged**